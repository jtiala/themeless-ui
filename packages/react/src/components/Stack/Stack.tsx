import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type StackProps = {
  /**
   * The direction to stack the elements.
   */
  direction?: "vertical" | "horizontal";

  /**
   * The elements to be stacked.
   */
  children?: ReactNode;
} & CommonComponentProps;

const stackClassName = cn("stack");
const verticalClassName = cn("stack-vertical");
const horizontalClassName = cn("stack-horizontal");

/**
 * Display a stack of items.
 */
export function Stack({
  direction = "vertical",
  children,
  testId,
}: StackProps) {
  const classNames = [
    stackClassName,
    direction === "vertical" ? verticalClassName : horizontalClassName,
  ].join(" ");

  return (
    <div className={classNames} data-testid={testId}>
      {children}
    </div>
  );
}
