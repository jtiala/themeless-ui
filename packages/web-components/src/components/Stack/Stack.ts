import style from "@themeless-ui/style/dist/stack.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { TUIComponent } from "../TUIComponent";

const stackClassName = cn("stack");
const verticalClassName = cn("stack-vertical");
const horizontalClassName = cn("stack-horizontal");

type Direction = "vertical" | "horizontal";

@customElement("tui-stack")
export class Stack extends TUIComponent {
  /**
   * The direction to stack the elements.
   *
   * @default vertical
   */
  @property()
  direction!: Direction;

  static styles = unsafeCSS(style);

  constructor() {
    super();

    this.direction = "vertical";
  }

  override render() {
    const classNames = [
      stackClassName,
      this.direction === "vertical" ? verticalClassName : horizontalClassName,
    ].join(" ");

    return html`
      <div
        class="${classNames}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </div>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-stack": Stack;
  }
}
