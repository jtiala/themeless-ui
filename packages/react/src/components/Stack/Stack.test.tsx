import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Stack } from "./Stack";

const defaultStack = (
  <Stack testId="stack">
    <div>1</div>
    <div>2</div>
    <div>3</div>
  </Stack>
);

describe("Stack", async () => {
  it("should render the stack", () => {
    expect(componentToJson(renderer.create(defaultStack))).toMatchSnapshot();

    render(defaultStack);

    const stack = screen.getByTestId("stack");
    const stackedElements = stack.getElementsByTagName("div");

    expect(stack).toBeInTheDocument();
    expect(stackedElements.length).toBe(3);
  });
});
