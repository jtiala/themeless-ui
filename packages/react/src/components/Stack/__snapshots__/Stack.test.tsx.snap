// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stack > should render the stack 1`] = `
<div
  className="tui__stack tui__stack-vertical"
  data-testid="stack"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;
